pkg/reconciler/apis/apiexportendpointsliceurls/apiexportendpointslice_controller.go

```diff
@@ -0,0 +1,496 @@
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAllAPIExportEndpointSlices(obj)
+		},
+	}))
+
+	return c, nil
+}
+
+// controller reconciles APIExportEndpointSlices. It ensures that the shard endpoints are populated
+// in the status of every APIExportEndpointSlices.
+type controller struct {
+	queue         workqueue.TypedRateLimitingInterface[string]
+	shardName     string
+	clusterClient kcpclientset.ClusterInterface
+
+	listShards                  func(selector labels.Selector) ([]*corev1alpha1.Shard, error)
+	listAPIExportEndpointSlices func() ([]*apisv1alpha1.APIExportEndpointSlice, error)
+	getAPIExportEndpointSlice   func(ctx context.Context, path logicalcluster.Path, name string) (*apisv1alpha1.APIExportEndpointSlice, error)
+	getAPIExport                func(path logicalcluster.Path, name string) (*apisv1alpha1.APIExport, error)
+	listAPIBindingsByAPIExport  func(apiexport *apisv1alpha1.APIExport) ([]*apisv1alpha1.APIBinding, error)
+
+	apiExportEndpointSliceClusterInformer       apisv1alpha1informers.APIExportEndpointSliceClusterInformer
+	globalApiExportEndpointSliceClusterInformer apisv1alpha1informers.APIExportEndpointSliceClusterInformer
+}
+
+func (c *controller) enqueueAPIExportEndpointSliceByAPIBinding(obj interface{}) {
+	if d, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+		obj = d.Obj
+	}
+	binding, ok := obj.(*apisv1alpha1.APIBinding)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("obj is supposed to be a APIBinding, but is %T", obj))
+		return
+	}
+
+	{ // local to shard
+		keys := sets.New[string]()
+		if path := logicalcluster.NewPath(binding.Spec.Reference.Export.Path); !path.Empty() {
+			pathKeys, err := c.apiExportEndpointSliceClusterInformer.Informer().GetIndexer().IndexKeys(indexAPIExportEndpointSliceByAPIExport, path.Join(binding.Spec.Reference.Export.Name).String())
+			if err != nil {
+				utilruntime.HandleError(err)
+				return
+			}
+			keys.Insert(pathKeys...)
+		}
+
+		for _, key := range sets.List[string](keys) {
+			slice, exists, err := c.apiExportEndpointSliceClusterInformer.Informer().GetIndexer().GetByKey(key)
+			if err != nil {
+				utilruntime.HandleError(err)
+				continue
+			} else if !exists {
+				continue
+			}
+			logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), obj.(*apisv1alpha1.APIBinding))
+			logging.WithQueueKey(logger, key).V(4).Info("queuing APIExportEndpointSlices because of consumer APIBinding")
+			c.enqueueAPIExportEndpointSlice(slice)
+		}
+	}
+	{
+		keys := sets.New[string]()
+		if path := logicalcluster.NewPath(binding.Spec.Reference.Export.Path); !path.Empty() {
+			pathKeys, err := c.globalApiExportEndpointSliceClusterInformer.Informer().GetIndexer().IndexKeys(indexAPIExportEndpointSliceByAPIExport, path.Join(binding.Spec.Reference.Export.Name).String())
+			if err != nil {
+				utilruntime.HandleError(err)
+				return
+			}
+			keys.Insert(pathKeys...)
+		}
+
+		for _, key := range sets.List[string](keys) {
```

Comment:
isn't `sets.List` ordering? We don't need an order.

Reply:
this pattern is being used all around the code. I think we either leave it or change everywhere.